DNSPolicy in a single cluster context (#30)
Kuadrant/architecture: rfcs/0000-single-cluster-dnspolicy.md

* Decouple the API entirely from OCM and multi cluster specific concepts.
* Simplify the DNS record structure created for a gateway listeners host for single cluster use.
* Improve the likelihood of adoption by creating an integration path for other kubernetes dns controllers such as [external-dns](https://github.com/kubernetes-sigs/external-dns).

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

The DNSPolicy can be used to target a Gateway in a single cluster context and will create dns records for each listener host in an appropriately configured external dns provider.
In this context the advanced `loadbalancing` configuration is unnecessary, and the resulting DNSRecord can be created mapping individual listener hosts to a single DNS A or CNAME record by using the `simple` strategy in the DNSPolicy.

**Example 1.** DNSPolicy using `simple` strategy

```yaml
apiVersion: kuadrant.io/v1alpha1
kind: DNSPolicy
metadata:
  name: prod-web
  namespace: my-gateways
spec:
  providerRef:
    name: my-route53-credentials
    namespace: my-gateways
    kind: Secret
  targetRef:
    name: prod-web
    group: gateway.networking.k8s.io
    kind: Gateway

>   strategy: simple

Reviewer: been thinking about this property. WDYT to changing it to be `routingStrategy` might convey a little better what it does?

```

```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: prod-web
  namespace: my-gateways
spec:
  gatewayClassName: istio
  listeners:
    - allowedRoutes:
        namespaces:
          from: All
      name: api
      hostname: "myapp.mn.hcpapps.net"
      port: 80
      protocol: HTTP
status:
  addresses:
    - type: IPAddress
      value: 172.31.200.0
```

In the example the `api` listener has a hostname `myapp.mn.hcpapps.net` that matches a hosted zone being managed by the provider referenced `my-route53-credentials` in the DNSPolicy. 
As the `simple` strategy is set in the DNSPolicy a DNSRecord resource with the following contents will be created:

```yaml
apiVersion: kuadrant.io/v1alpha1
kind: DNSRecord
metadata:
  name: prod-web-api
  namespace: my-gateways
spec:
  providerRef:
    name: my-route53-credentials
    namespace: my-gateways
  endpoints:
    - dnsName: myapp.mn.hcpapps.net
      recordTTL: 60
      recordType: A
      targets:

>         - 172.31.200.0

Reviewer: while this is focused around single cluster, do you think it is worth mentioning that this new `simple` strategy will work in a MC and will create a multi-value record?

Author: Yes, will add an example

```

> The `providerRef` is included in the DNSRecord to allow the MGC dns record controller to load the appropriate provider configuration during reconciliation and create the DNS records in the dns provider service e.g. route 53, by default the provider `Kind` is a secret.

Reviewer: as we are moving to a policy controller may be remove MGC?
```suggestion
The `providerRef` is included in the DNSRecord to allow the dns record controller to load the appropriate provider configuration during reconciliation and create the DNS records in the dns provider service e.g. route 53, by default the provider `Kind` is a secret.
```

**Example 2.** DNSPolicy using `simple` strategy with external dns provider

```yaml
apiVersion: kuadrant.io/v1alpha1
kind: DNSPolicy
metadata:
  name: prod-web
  namespace: my-gateways
spec:
  providerRef:
    name: external-dns
    kind: ExternalDNS
  targetRef:
    name: prod-web
    group: gateway.networking.k8s.io
    kind: Gateway
  strategy: simple
```

> In ths example if the DNSPolicy was attached to the same gateway described in example 1, the same DNSRecord would also be created but MGC would not reconcile it. In this scenario it is expected that an external controller is being used to manage the reconciliation of the DNSRecord resources such as [external-dns](https://github.com/kubernetes-sigs/external-dns).

Reviewer: ```suggestion
In ths example if the DNSPolicy was attached to the same gateway described in example 1, the same DNSRecord would also be created but the `DNSPolicy`  `DNSRecord` controller would not reconcile it. In this scenario it is expected that an external controller is being used to manage the reconciliation of the DNSRecord resources such as [external-dns](https://github.com/kubernetes-sigs/external-dns).
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## API Updates

DNSPolicy:

- new providerRef field `spec.providerRef`

> - - new strategy field `spec.strategy` 

Reviewer: ```suggestion
- new strategy field `spec.strategy` 
```

DNSRecord:
 
- `spec.managedZone` replaced with `spec.providerRef`

> - - new zoneID field `spec.zoneID`

Reviewer: ```suggestion
- new zoneID field `spec.zoneID`
```

ManagedZone:

- `spec.dnsProviderSecretRef` replaced with `spec.providerRef`

### DNSPolicy.spec.providerRef

The `providerRef` field is mandatory and contains a reference to a resource that shows how DNSRecords will be reconciled.
    - `spec.providerRef.name` - name of the provider resource

>     - `spec.providerRef.kind` - kind of resource, can be anything but only `Secret` or `Managedzone` will be reconciled by MGC.

Reviewer: ```suggestion
    - `spec.providerRef.kind` - kind of resource, can be anything but only `Secret` or `Managedzone` will be reconciled by the `DNSPolicy` controller.
```

> A kind of type `Secret` will be managed by MGC, and a secret in the dns policies namespace with the given name must exist. The expected contents of the secrets data is comparable to the `dnsProviderSecretRef` used by ManageZones.

Reviewer: ```suggestion
A kind of type `Secret` will be managed by the `DNSPolicy` controller, and a secret in the dns policies namespace with the given name must exist. The expected contents of the secrets data is comparable to the `dnsProviderSecretRef` used by ManageZones.
```

```yaml
apiVersion: v1
kind: Secret
metadata:

>   name: mgc-aws-credentials

Reviewer: ```suggestion
  name: aws-credentials
```

type: kuadrant.io/aws
data:
  AWS_ACCESS_KEY_ID: "foo"
  AWS_SECRET_ACCESS_KEY: "bar"
  CONFIG:
    zoneIDFilter:
      - Z04114632NOABXYWH93QUl
```

The `CONFIG` section of the secrets data will be added to allow provider specific configuration to be stored alongside the providers credentials and can be used during the instantiation of the provider client, and during any provider operations.
The above for example would use the `zoneIDFilter` value to limit what hosted zones this provider is allowed to update.

> A kind of type `ManagedZone` will be managed by MGC, and a ManagedZone in the dns policies namespace with the given name must exist. 

Reviewer: ```suggestion
A kind of type `ManagedZone` will be managed by the `DNSPolicy` controller, and a ManagedZone in the dns policies namespace with the given name must exist. 
```

> A kind of any other type e.g. `ExternalDNS` informs MGC that it should not reconcile any resources referencing it. All fields of `providerRef` will be ignored by MGC and can be set to anything, however since the `providerRef` will still be copied over to any DNSRecord resources created by the policy controller the values may still be given meaning to that external DNSRecord reconciler.

Reviewer: ```suggestion
A kind of any other type e.g. `ExternalDNS` informs the `DNSPolicy` controller that it should not reconcile any resources referencing it. All fields of `providerRef` will be ignored by the `DNSPolicy` controller and can be set to anything, however since the `providerRef` will still be copied over to any DNSRecord resources created by the policy controller the values may still be given meaning to that external DNSRecord reconciler.
```

### DNSPolicy.spec.strategy[simple|loadalanced]

The `strategy` field is mandatory and dictates what kind of dns record structure the policy will create. Two strategy options are allowed `simple` or `loadbalanced`.

A reconciliation of DNSPolicy processes the target gateway and creates a DNSRecord per listener that is supported by the currently configured provider(hostname matches the hosted zones accessible with the credentials and config). The strategy used will determine the contents of the DNSRecord resources Endpoints array.

#### Simple

```yaml
apiVersion: kuadrant.io/v1alpha1
kind: DNSRecord
spec:
  providerRef:
    name: my-route53-credentials
    namespace: my-gateways
    kind: Secret
  endpoints:
    - dnsName: myapp.mn.hcpapps.net
      recordTTL: 60
      recordType: A
      targets:

>         - 172.31.200.0

Reviewer 2: How will the `simple` strategy look for Multi-cluster deployment? `A` record for each cluster

Reviewer: No in an hub it would create a multi value A record

Author: @pehala Added example https://github.com/Kuadrant/architecture/pull/30/commits/3fef6f3e4cb9ac8587b18f89e5a8f5c6e1104ea1

### DNSRecord.spec.zoneID

The `zoneID` field is mandatory and contains the provider specific id of the hosted zone that this record should be published into. 

The DNSRecord reconciliation will use this zone when creating/updating or deleting endpoints for this record set. 

The `zoneID` should not change after being selected during initial creation and as such will be marked as immutable.

### ManagedZone.spec.providerRef

More details of `providerRef` found in [DNSPolicy.spec.providerRef](#dnspolicyspecproviderref)

Replaces the existing `dnsProviderSecretRef` for consistency with other resources that require a provider reference (DNSRecord and DNSPolicy). 

In the case of a ManagedZone a providerRef kind of type `ManagedZone` will not be allowed and will be rejected during create/update.  

# Prior art
[prior-art]: #prior-art

[ExternalDNS](https://github.com/kubernetes-sigs/external-dns)

* Uses annotations on the target Gateway as opposed to a proper API.
* Requires access to the HTTP route resources.
* Supports only a single provider per external dns instance.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

> When a provider is configured using a kind not supported by MGC e.g. `ExternalDNS` we will be relying on an external controller to correctly update the status of any DNSRecord resources created by our policy. This may have a negative impact on our ability to correctly report status back to the target resource.

Reviewer: ```suggestion
When a provider is configured using a kind not supported by the `DNSPolicy` controller e.g. `ExternalDNS` we will be relying on an external controller to correctly update the status of any DNSRecord resources created by our policy. This may have a negative impact on our ability to correctly report status back to the target resource.
```

# RFC Template

- Feature Name: `single-cluster-dnspolicy`
- Start Date: 2023-10-09
- RFC PR: [Kuadrant/architecture#0000](https://github.com/Kuadrant/architecture/pull/0000)
- Issue tracking: [Kuadrant/architecture#0000](https://github.com/Kuadrant/architecture/issues/0000)

# Summary
[summary]: #summary

> Proposal for changes to the `DNSPolicy` API to allow it to work more effectively in a single cluster context.

Reviewer: ```suggestion
Proposal for changes to the `DNSPolicy` API to allow it to provide a simple dns strategy in a single cluster removing the need for the complex DNS structure required in a multi-cluster environment and in doing so allow use of popular dns integrators such as external-dns .
```

# Motivation
[motivation]: #motivation

The [`DNSPolicy`](https://github.com/Kuadrant/multicluster-gateway-controller/blob/main/pkg/apis/v1alpha1/dnspolicy_types.go) API (v1alpha1), was implemented as part of our multi cluster gateway offering using OCM and as such the design and implementation were influenced heavily by how we want multi cluster dns to work.

* Decouple the API entirely from OCM and multi cluster specific concepts.

> * Simplify the DNS record structure created for a gateway listeners host.

Reviewer: ```suggestion
* Simplify the DNS record structure created for a gateway listeners host for single cluster use.
```

    namespace: my-gateways
    kind: Secret
  targetRef:
    name: prod-web
    group: gateway.networking.k8s.io
    kind: Gateway
  strategy: simple
```

```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: prod-web
  namespace: my-gateways
spec:
  gatewayClassName: istio
  listeners:
    - allowedRoutes:
        namespaces:
          from: All
      name: api
      hostname: "myapp.mn.hcpapps.net"
      port: 80
      protocol: HTTP
status:
  addresses:
    - type: IPAddress
      value: 172.31.200.0
```

In the example the `api` listener has a hostname `myapp.mn.hcpapps.net` that matches a hosted zone being managed by the provider referenced `my-route53-credentials` in the DNSPolicy. 
As the `simple` strategy is set in the DNSPolicy a DNSRecord resource with the following contents will be created:

```yaml
apiVersion: kuadrant.io/v1alpha1
kind: DNSRecord
metadata:
  name: prod-web-api
  namespace: my-gateways
spec:
  providerRef:
    name: my-route53-credentials
    namespace: my-gateways
  endpoints:
    - dnsName: myapp.mn.hcpapps.net
      recordTTL: 60
      recordType: A
      targets:
        - 172.31.200.0
```

> The `providerRef` is included in the DNSRecord to allow the MGC dns record controller to load the appropriate provider during reconciliation and create the DNS records in the dns provider service e.g. route 53.

Reviewer: ```suggestion
The `providerRef` is included in the DNSRecord to allow the MGC dns record controller to load the appropriate provider configuration during reconciliation and create the DNS records in the dns provider service e.g. route 53. By default the provider `Kind` is a secret
```

```yaml
apiVersion: kuadrant.io/v1alpha1
kind: DNSPolicy
metadata:
  name: prod-web
  namespace: my-gateways
spec:
  providerRef:
    name: external-dns
    kind: ExternalDNS
  targetRef:
    name: prod-web
    group: gateway.networking.k8s.io
    kind: Gateway
  strategy: simple
```

In ths example if the DNSPolicy was attached to the same gateway described in example 1, the same DNSRecord would also be created but MGC would not reconcile it. In this scenario it is expected that an external controller is being used to manage the reconciliation of the DNSRecord resources such as [external-dns](https://github.com/kubernetes-sigs/external-dns).

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## API Updates

DNSPolicy:

* - new providerRef field `spec.providerRef` 
* - new strategy field `spec.strategy` 

DNSRecord:

- `spec.managedZone` replaced with `spec.providerRef`

### DNSPolicy.spec.providerRef

The `providerRef` field is mandatory and contains a reference to a resource that shows how DNSRecords will be reconciled.
    * - `spec.providerRef.name` - name of the provider resource

>     * - `spec.providerRef.namespace` - namespace of provider resource

Reviewer: should we allow cross namespace? I am thinking we don't at least initially. By specifying a namespace, we allow someone in one ns to "guess" and grant access to a secret in another ns.

Author: I have no objection to removing it. Was carried over form the ManagedZone reference that already existed, and already allows this. I do however agree that just looking for a secret in the current namespace is probably safer. Will update accordingly.

>     * - `spec.providerRef.kind` - kind of resource, can be anything but only `Secret` will be reconciled by MGC.

Reviewer: is it fair to say that any value other than the default `secret` will turn off DNSRecord reconciling

Reviewer: NVM I see it below

> A kind of type `Secret` will be managed by MGC, and a secret in the given namespace with the given name must exist. The expected contents of the secrets data is comparable to the `dnsProviderSecretRef` used by ManageZones. 

Reviewer: should we align these? Prob move managedzone to use ```providerRef``` also? its  not part of this obv, could just be reflected in an issue

Author: Yep, this will be renamed, but in this doc i guess we are talking about the current version of the code which has it as `dnsProviderSecretRef`.

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: mgc-aws-credentials
  namespace: multi-cluster-gateways
type: kuadrant.io/aws
data:
  AWS_ACCESS_KEY_ID: "foo"
  AWS_SECRET_ACCESS_KEY: "bar"
  CONFIG:
    zoneIDFilter:
      - Z04114632NOABXYWH93QUl
```

> The `CONFIG` section of the secrets data will be added to allow provider specific configuration to be stored alongside the providers credentials and can be used during the instantiation of the provider client, and during any provider operations.

Reviewer: I wonder do we want to call out that this filter will decide whether or not a DNSRecord can be reconciled into the DNSProvider. If no appropriate zone is found, then we will fail to reconcile the the dnsrecord. Should there be some form of 
```listenerStatus``` on the DNSPolicy added that reflects whether the DNSPolicy is ready for that listener?
Also thikn that we should probably reflect some form of listener status condition on the gateway

Reviewer: We might want a section in here around how we want status to look in the DNSPolicy as part of this?

        - name: weight
          value: "120"
      recordTTL: 60
      recordType: CNAME
      setIdentifier: lrnse3.lb-4ej5le.myapp.mn.hcpapps.net
      targets:
        - lrnse3.lb-4ej5le.myapp.mn.hcpapps.net
    - dnsName: lrnse3.lb-4ej5le.myapp.mn.hcpapps.net
      recordTTL: 60
      recordType: A
      targets:
        - 172.31.200.0
```

Note: `LoadBalanced` is the current default for DNSPolicy and more details about it can be found in the original [DNSPolicy rfc](https://github.com/Kuadrant/architecture/blob/main/rfcs/0003-dns-policy.md) 

### DNSRecord.spec.providerRef

The DNSRecord API is updated to remove the `managedZone` reference in favour of directly referencing the `providerRef` credentials instead. The DNSRecord reconciliation will be unchanged except for loading the provider client from `providerRef` credentials.

The DNSPolicy reconciliation will be updated to remove the requirement for a ManagedZone resource to be created before a DNSPolicy can create dns records for it, instead it will be replaced in favour of just listing available zones directly in the currently configured dns provider. If no matching zone is found, no DNSRecord will be created.

> There is a potential for a DNSRecord to be created successfully, but then a provider updated to remove access. In this case it is the responsibility of the DNSPolicy controller to remove DNSRecords it created that can no longed be reconciled into the configured provider. 

Reviewer: I wonder would it be better to add something to the status of these rather than remove them?

Author: I'm not sure it (the policy controller) should directly update the status of the DNSRecord, the DNSRecord reconciliation will already update that anyway if it can't access the provider for whatever reason. In that scenario it comes back to aggregating the status of the policy created resources into the policy status and/or the gateway status (Separate proposal we talked about).

I agree though it probably shouldn't just remove it though, will update that.

  listeners:
    - allowedRoutes:
        namespaces:
          from: All
      name: api
      hostname: "myapp.mn.hcpapps.net"
      port: 80
      protocol: HTTP
status:
  addresses:
    - type: kuadrant.io/MultiClusterIPAddress
      value: 172.31.200.0
    - type: kuadrant.io/MultiClusterIPAddress
      value: 172.31.201.0
```

Similar to example 1, except here the Gateway is a multi cluster gateway that has had its status updated by the `Gateway` controller to include `kuadrant.io/MultiClusterIPAddress` type addresses.
As the `simple` routing strategy is set in the DNSPolicy a DNSRecord resource with the following contents will be created:

```yaml
apiVersion: kuadrant.io/v1alpha1
kind: DNSRecord
metadata:
  name: prod-web-api
  namespace: my-gateways
spec:
  providerRef:
    name: my-route53-credentials
    namespace: my-gateways
  endpoints:
    - dnsName: myapp.mn.hcpapps.net
      recordTTL: 60
      recordType: A
      targets:
        - 172.31.200.0
        - 172.31.201.0
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

## API Updates

DNSPolicy:

- new providerRef field `spec.providerRef`

> - new routingStrategy field `spec.routingStrategy` 

Reviewer: ```suggestion
- new routingStrategy field `spec.routingStrategy` 
- new api version `v1alpha2`
```

DNSRecord:
 
- `spec.managedZone` replaced with `spec.providerRef`
- new zoneID field `spec.zoneID`

ManagedZone:

> - `spec.dnsProviderSecretRef` replaced with `spec.providerRef`

Reviewer: ```suggestion
- `spec.dnsProviderSecretRef` replaced with `spec.providerRef`
- new API version `v1alpha2`
```

Author: https://github.com/Kuadrant/architecture/pull/30/commits/7eab96b667abae16c40ebc538891bbb9ad6ff9a9
